### ISSUE TITLE
Body collect seems to have a memory leak when body is large.

### ISSUE THREAD

**Issue by kevinzhow** [vapor/vapor#3223]
### Describe the issue

Body collect seems to have a memory leak. I set max('30mb') in body collect for testing, and after requesting an empty handler, the memory was not released in the end.

### Vapor version

4.102.1

### Operating system and version

Ubuntu Linux 24.04

### Swift version

Swift 5.10

### Steps to reproduce

1. Clone https://github.com/kevinzhow/body-collect-leak
2. run `docker compose up`
3. run `ab -n 1024 -c 10 -T text/plain -p large_file.txt http://127.0.0.1:8080/file/upload`
4. Open a new terminal and run `docker stats`

### Outcome

After the requests are finished, the memory will never be released.

<img width="939" alt="image" src="https://github.com/user-attachments/assets/7e12fa8b-daa4-484f-be3f-0a679778b00b">

**Comment by 0xTim** [comment]
Are you running with Jemalloc? Regular malloc has a habit of not releasing memory back to the OS

**Comment by kevinzhow** [comment]
> Are you running with Jemalloc? Regular malloc has a habit of not releasing memory back to the OS

Yes, i am using Jemalloc, it was enabled in the Dockerfile.

**Comment by mkll** [comment]
> Yes, i am using Jemalloc

Jemalloc is not known for freeing up memory immediately, but for not allowing it to grow indefinitely, which can eventually lead to an application crash. Observe it for at least a day, then you can draw some conclusions. It's too early for that now.

**Comment by kevinzhow** [comment]
@mkll 

Yes, I have been monitoring this issue for 3 days. As shown in the chart below, even with Jemalloc, the memory usage keeps increasing. The drops you see are due to my manual server restarts. This issue doesn’t occur on macOS, but it’s easily reproducible on Linux.

<img width="1539" alt="image" src="https://github.com/user-attachments/assets/6ca5bbc8-e004-4dd9-8cc7-f2ff1fd9798a">

**Comment by mkll** [comment]
@kevinzhow Looking at the graph, it is impossible to draw any clear conclusion. The gradual increase in memory may well be part of the memory management policy of jemalloc.

Also, the fact that the effect only occurs on Linux indirectly suggests that the cause lies outside Swift's memory management?

P.S. Did you read https://github.com/vapor/vapor/issues/2493?

**Comment by mkll** [comment]
By the way, in Debian, starting with version 12, the standard allocator was replaced with a custom one, but not jemalloc, but Microsoft's [mimalloc](https://github.com/microsoft/mimalloc). It might be worth giving it a try.

**Comment by kevinzhow** [comment]
> u

Yes, I have read this issue. In fact, I was previously using a 1GB t3a.micro instance, but this issue caused my server to crash when the memory hit 100%. So, I switched to a t4g.small instance with 2GB of memory, which resulted in the chart above. Although I could move the logic for processing this large file out of Vapor, I’m sharing this in case anyone has better insights.

**Comment by kevinzhow** [comment]
> @kevinzhow Looking at the graph, it is impossible to draw any clear conclusion. The gradual increase in memory may well be part of the memory management policy of jemalloc.
> 
> Also, the fact that the effect only occurs on Linux indirectly suggests that the cause lies outside Swift's memory management.
> 
> P.S. Did you read #2493?

Yes, I have read this issue. In fact, I was previously using a 1GB t3a.micro instance, but this issue caused my server to crash when the memory hit 100%. So, I switched to a t4g.small instance with 2GB of memory, which resulted in the chart above. Although I could move the logic for processing this large file out of Vapor, I’m sharing this in case anyone has better insights.

**Comment by kevinzhow** [comment]
> By the way, in Debian, starting with version 12, the standard allocator was replaced with a custom one, not jemalloc, but Microsoft's [mimalloc](https://github.com/microsoft/mimalloc). It might be worth giving it a try.

Thank you for the suggestion, I’ll give it a try!

**Comment by mkll** [comment]
@kevinzhow It would be interesting to hear about the results of your attempts later.

**Comment by kevinzhow** [comment]
@mkll After switching to mimalloc, the memory usage drops back to 2.8% within a minute. However, with jemalloc, some memory remains unreleased for a long time, eventually stabilizing at 7.2%.

# mimalloc cold start
<img width="734" alt="image" src="https://github.com/user-attachments/assets/47bc20ae-4645-4481-8c71-e43c1427e6e1">

# mimalloc after 100 requests
<img width="650" alt="image" src="https://github.com/user-attachments/assets/bc1265d3-2923-4932-b66e-cccb2163c977">

# jemalloc cold start
<img width="629" alt="image" src="https://github.com/user-attachments/assets/1e30be27-a0ff-461d-9762-47a9b7b1b558">

# jemalloc after 100 requests
<img width="655" alt="image" src="https://github.com/user-attachments/assets/a4296e93-c2a8-469b-9738-5e85907ef285">

**Comment by mkll** [comment]
@kevinzhow Very interesting and useful experiment, thank you! Until now I have not seen a direct comparison of jemalloc and mimalloc.

**Comment by kevinzhow** [comment]
> @kevinzhow Very interesting and useful experiment, thank you! Until now I have not seen a direct comparison of jemalloc and mimalloc.

Thank you for suggesting mimalloc :) I’ve now switched it to the server, and I’ll report back on its performance after a few hours.

**Comment by kevinzhow** [comment]
@mkll Turns out still the same
<img width="796" alt="image" src="https://github.com/user-attachments/assets/58226f12-a0da-4402-9812-29e446df2193">

**Comment by miroslavkovac** [comment]
@ptoffy is the #3232 expected to also fix this memory issue? I have also observed this linear ever-increasing memory allocation, similar to what @kevinzhow reported.
